runtime/lib/capi/RuntimeCAPI.cpp

```diff
@@ -242,7 +242,7 @@ QirArray *__quantum__rt__qubit_allocate_array(int64_t num_qubits)
 
     // I don't like this copying.
     std::vector<QubitIdType> *qubit_vector_ptr =
-        new std::vector<QubitIdType>(qubit_vector.begin(), qubit_vector.end());
+        new std::vector<QubitIdType>(qubit_vector.begin(), qubit_vector.begin() + num_qubits);
```

Comment:
@maliasadi do you know if this is correct? I could imagine it instead being. 

```c++
        auto size = qubit_vector.size();
        new std::vector<QubitIdType>(qubit_vector.begin() + size - num_qubits, qubit_vector.end);
```

The behaviour in the commit is the one present in the run time at the moment.

Reply:
Are you trying to make this work with multiple allocations? I think in either case the `AllocateQubits` function should only return a vector of the newly allocated qubit ids.

Therefor,
```cpp
new std::vector<QubitIdType>(qubit_vector.begin(), qubit_vector.end());
```
should be correct (and equivalent to `(qubit_vector.begin(), qubit_vector.begin() + num_qubits)`, could be an assertion).

Reply:
This is incorrect. See the previous commit and the failure in kokkos: https://github.com/PennyLaneAI/catalyst/actions/runs/7612799519/job/20731193678

Reply:
Might be a bug in that case?

Reply:
We can see the behaviour deviates from the Lightning Qubit implementation, which actually returns an array of the right size:

```cpp
auto LightningSimulator::AllocateQubits(size_t num_qubits) -> std::vector<QubitIdType>
{
    if (num_qubits == 0U) {
        return {};
    }

    // at the first call when num_qubits == 0
    if (this->GetNumQubits() == 0U) {
        this->device_sv = std::make_unique<StateVectorT>(num_qubits);
        return this->qubit_manager.AllocateRange(0, num_qubits);
    }

    std::vector<QubitIdType> result(num_qubits);
    std::generate_n(result.begin(), num_qubits, [this]() { return AllocateQubit(); });
    return result;
}
```

Reply:
I opened a bug ticket in shortcut for it ([vector size](https://app.shortcut.com/xanaduai/story/55283) and [state loss](https://app.shortcut.com/xanaduai/story/55284)), but I don't think we should merge code in this PR that corresponds to the buggy behaviour.

The `qubit_vector.begin(), qubit_vector.begin() + num_qubits` version works around the bug, while still being correct, although `qubit_vector.begin(), qubit_vector.end()` is more "intuitively" correct. I think adding an assertion that they are the same would be good idea and will catch similar issues with the same or other devices in the future.

Reply:
yep, there is indeed a bug in `AllocateRange`. (Check this PR for the fix #473)

> So it creates a new vector, reserves a size that is the total number of qubits (size = new_num_qubits), then starts pushing back qubits from the first new index (start_idx = cur_num_qubits), but pushes back a number of times equal to the total number of qubits?

This is more like an implementation assumption we had in mind while developing the `QubitManager`. Although, `this->qubits_map` in the manager hasn't been really used anywhere, we should see some application of this with the dynamic qubit allocations/deallocations support in Catalyst.

> I also just noticed that it seems to be resetting the statevector? In that case I don't know why it would keep around the previous qubits since they will be overwritten at this point.

This method extends the number of allocated qubits. for devices without dynamic allocations/deallocations, we decided to reset the state while still increasing the number of qubits, so that the qubit manager can map device and host qubit indices correctly after re-allocation. Looking back to this support, I think we should raise a runtime error instead:
``` c++
auto OpenQasmDevice::AllocateQubits(size_t num_qubits) -> std::vector<QubitIdType>
{
    if (num_qubits == 0U) {
        return {};
    }

    const size_t cur_num_qubits = builder->getNumQubits();
    RT_FAIL_IF(cur_num_qubits, "Partial qubits allocation is not supported by OpenQasmDevice");

    const size_t new_num_qubits = cur_num_qubits + num_qubits;
    if (cur_num_qubits) {
        builder = std::make_unique<OpenQasm::OpenQasmBuilder>();
    }

    return qubit_manager.AllocateRange(cur_num_qubits, num_qubits);
}

```

Reply:
Resolving this conversation. Thanks for the discussion!

Reply:
What is the resolution you are implementing? I would strongly urge not to keep the behaviour that enabled the bug to begin with.

Reply:
Ah sorry. Sure, I will just copy the values returned by AllocateQubits. :sweat_smile:

Reply:
As you've mentioned here:

> Therefor,
>
> `new std::vector<QubitIdType>(qubit_vector.begin(), qubit_vector.end());`
>
> should be correct (and equivalent to `(qubit_vector.begin(), qubit_vector.begin() + num_qubits)`, could be an assertion)

Which it wasn't due to AllocateQubits. But since it has been fixed, there is no problem now, I will just change it to `new std::vector<QubitIdType>(qubit_vector.begin(), qubit_vector.end());`